codec/src/main/java/io/netty/handler/codec/compression/Decompressor.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+
+/**
+ * Decompressor that takes care of decompress some input.
+ */
+public interface Decompressor extends AutoCloseable {
+    /**
+     * This method will read from the input {@link ByteBuf} and decompress into a new {@link ByteBuf} that will be
+     * allocated (if needed) from the {@link ByteBufAllocator}. If there is not enough readable data in the
+     * {@link ByteBuf} to process it will return {@code null}.
+     *
+     * This method should be called in a loop as long:
+     *
+     * <li>
+     *     <ul>{@link #isFinished()} is {@code false}</ul>
+     *     <ul>something was read from the {@code input}</ul>
+     *     <ul>something was returned</ul>
+     * </li>
+     * Otherwise this method should be called again once there is more data in the input buffer.
+     *
+     * @param input         the {@link ByteBuf} that contains the data to be decompressed.
+     * @param allocator     the {@link ByteBufAllocator} that is used to allocate a new buffer (if needed) to write the
+     *                      decompressed bytes too.
+     * @return              the {@link ByteBuf} that contains the decompressed data. The caller of this method takes
+     *                      ownership of the buffer. The return value will be {@code null} if there is not enough data
+     *                      readable in the input to make any progress. In this case the user should call it again once
+     *                      there is more data ready to be consumed.
+     * @throws DecompressionException   thrown if an decompression error was encountered.
+     */
+    ByteBuf decompress(ByteBuf input, ByteBufAllocator allocator) throws DecompressionException;
+
+    /**
+     * Returns {@code} true if the decompressor was closed. This might be because someone explicit called
```

Comment:
Perhaps call this `isClosed()` since we do not have any `finish` semantics in decompressors. 

Also I think we should distinguish between internal closure and external closure of compressor and decompressor. Having a resource which can close itself is confusing as it implies that it can happen spuriously and the caller may have to check for `isFinished()` before calling `decompress()`. I think having the ownership of lifecycle with the owner of the object is clearer, i.e. `isFinished()` will only return `true` if `finish()` has been called. Which subsequently brings the question that whether we need this method if the ownership is clear 😄